README.md
> This is a fork to the now unmaintained project [react-inline-edit](https://github.com/kaivi/ReactInlineEdit).
> Only minimal maintainance will be provided, and __no new features will be accepted__. 

# Inline Edit Component for React

[![npm version](https://badge.fury.io/js/react-edit-inplace.svg)](https://badge.fury.io/js/react-edit-inplace) 

This is a simple React component for in-place text editing. It turns into an `<input />` when focused, and tries to validate and save input on <kbd>Enter</kbd> or `blur`. <kbd>Esc</kbd> works as well for cancelling.

![Example animation gif](http://i.imgur.com/8vig5m1.gif)

Watch a [demo](http://htmlpreview.github.io/?https://github.com/dkwingsmt/ReactInlineEdit/blob/master/demo/index.html), then check out [demo/index.jsx](demo/index.jsx) for a quick example.

### Installation

`npm install --save react-edit-inplace`

### Required props
- `text`:`string` initial text
- `paramName`:`string` name of the parameter to be returned to `change` function
- `change`:`function` function to call when new text is changed and validated, it will receive `{paramName: value}`

### Optional props
- `className`:_string_ CSS class name
- `activeClassName`:_string_ CSS class replacement for when in edit mode
- `validate`:_function_ boolean function for custom validation, using this overrides the two props below
- `minLength`:_number_ minimum text length, **default** `1`
- `maxLength`:_number_ maximum text length, **default** `256`
- `editingElement`:_string_ element name to use when in edit mode (DOM must have `value` property) **default** `input`
- `staticElement`:_string_ element name for displaying data **default** `span`
- `editing`:_boolean_ If true, element will be in edit mode
- `tabIndex`:_number_ tab index used for focusing with TAB key **default** `0`
- `stopPropagation`:_boolean_ If true, the event onClick will not be further propagated.

### Usage example
```javascript
import React from 'react';
import InlineEdit from 'react-edit-inplace';

class MyParentComponent extends React.Component {

    constructor(props){
      super(props);
      this.dataChanged = this.dataChanged.bind(this);
      this.state = {
        message: 'ReactInline demo'
      }
    }

    dataChanged(data) {
        // data = { description: "New validated text comes here" }
        // Update your model from here
        console.log(data)
        this.setState({...data})
    }

    customValidateText(text) {
      return (text.length > 0 && text.length < 64);
    }

    render() {
        return (<div>
            <h2>{this.state.message}</h2>
            <span>Edit me: </span>
            <InlineEdit
              validate={this.customValidateText}
              activeClassName="editing"
              text={this.state.message}
              paramName="message"
              change={this.dataChanged}
              style={{
                backgroundColor: 'yellow',
                minWidth: 150,
                display: 'inline-block',
                margin: 0,
                padding: 0,
                fontSize: 15,
                outline: 0,
                border: 0
              }}
            />
        </div>)
    }
}
```


CHANGELOG.md
1.1.1 / 2017-10-27
==================
* Update peer dependencies to allow React 15.6.x

1.1.0 / 2017-10-16
==================
* Move `react` and `prop-types` from `dependencies` to `peerDependencies`.

1.0.9 / 2017-10-16
==================
* Republish the package under name `react-edit-inplace`.
* Merged PR [#35](https://github.com/kaivi/ReactInlineEdit/pull/35)
  - Update propTypes usage to allow React 16.

1.0.8 / 2016-11-07
==================
* Merged PR [#26](https://github.com/kaivi/ReactInlineEdit/pull/26)
  - Remove onReturn props From Element

1.0.7 / 2016-04-28
==================

* Merged PR [#16](https://github.com/kaivi/ReactInlineEdit/pull/16)
  - Implement option to disable editing
* Merged PR [#17](https://github.com/kaivi/ReactInlineEdit/pull/16)
  - Update peer dependencies to allow React 15.x.x

1.0.6 / 2016-01-25
==================

* Merged PR [#11](https://github.com/kaivi/ReactInlineEdit/pull/11)
  - fixed a bug introduced in [#6](https://github.com/kaivi/ReactInlineEdit/pull/6)

1.0.5 / 2016-01-19
==================

* Merged PR [#9](https://github.com/kaivi/ReactInlineEdit/pull/9)
  - consistent style (and basic lint stuff)
  - spacing
  - spaces after if
  - let instead of var
  - dangling commas
  - lower case `functionName`
  - life-cycle methods before others, followed by render.
  - class properties (since stage 0 is being utilized)
  - other initialization stuff in `componentWillMount` as opposed to `constructor`

* Upgraded peerDependency to React 0.14.6 PR [#10](https://github.com/kaivi/ReactInlineEdit/pull/10)

1.0.4 / 2016-01-18
==================

* Added `editingElement` and `staticElement` props, so one can pass custom component names: [\#5](https://github.com/kaivi/ReactInlineEdit/issues/5)
* Added proper Babel v6.4 dependency to `devDependencies`, so it just builds on `npm install` with OS X
* Added a small demo under `demo/` directory
* Added tab stops via `tabIndex` property on `staticElement`
* Added custom `style` prop to pass styles, obviously
* When `text` prop changes, the component is now re-rendered: [\#6](https://github.com/kaivi/ReactInlineEdit/pull/6)
* Upgraded React peer dependency, so it matches React of any `v0.x.x` version
* Updated `LICENSE` to MIT, as in `package.json`


index.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function selectInputText(element) {
    element.setSelectionRange(0, element.value.length);
}

var InlineEdit = function (_React$Component) {
    _inherits(InlineEdit, _React$Component);

    function InlineEdit() {
        var _Object$getPrototypeO;

        var _temp, _this, _ret;

        _classCallCheck(this, InlineEdit);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(InlineEdit)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.state = {
            editing: _this.props.editing,
            text: _this.props.text,
            minLength: _this.props.minLength,
            maxLength: _this.props.maxLength
        }, _this.startEditing = function (e) {
            if (_this.props.stopPropagation) {
                e.stopPropagation();
            }
            _this.setState({ editing: true, text: _this.props.text });
        }, _this.finishEditing = function () {
            if (_this.isInputValid(_this.state.text) && _this.props.text != _this.state.text) {
                _this.commitEditing();
            } else if (_this.props.text === _this.state.text || !_this.isInputValid(_this.state.text)) {
                _this.cancelEditing();
            }
        }, _this.cancelEditing = function () {
            _this.setState({ editing: false, text: _this.props.text });
        }, _this.commitEditing = function () {
            _this.setState({ editing: false, text: _this.state.text });
            var newProp = {};
            newProp[_this.props.paramName] = _this.state.text;
            _this.props.change(newProp);
        }, _this.clickWhenEditing = function (e) {
            if (_this.props.stopPropagation) {
                e.stopPropagation();
            }
        }, _this.isInputValid = function (text) {
            return text.length >= _this.state.minLength && text.length <= _this.state.maxLength;
        }, _this.keyDown = function (event) {
            if (event.keyCode === 13) {
                _this.finishEditing();
            } else if (event.keyCode === 27) {
                _this.cancelEditing();
            }
        }, _this.textChanged = function (event) {
            _this.setState({
                text: event.target.value.trim()
            });
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(InlineEdit, [{
        key: 'componentWillMount',
        value: function componentWillMount() {
            this.isInputValid = this.props.validate || this.isInputValid;
            // Warn about deprecated elements
            if (this.props.element) {
                console.warn('`element` prop is deprecated: instead pass editingElement or staticElement to InlineEdit component');
            }
        }
    }, {
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            var isTextChanged = nextProps.text !== this.props.text;
            var isEditingChanged = nextProps.editing !== this.props.editing;
            var nextState = {};
            if (isTextChanged) {
                nextState.text = nextProps.text;
            }
            if (isEditingChanged) {
                nextState.editing = nextProps.editing;
            }
            if (isTextChanged || isEditingChanged) {
                this.setState(nextState);
            }
        }
    }, {
        key: 'componentDidUpdate',
        value: function componentDidUpdate(prevProps, prevState) {
            var inputElem = _reactDom2.default.findDOMNode(this.refs.input);
            if (this.state.editing && !prevState.editing) {
                inputElem.focus();
                selectInputText(inputElem);
            } else if (this.state.editing && prevProps.text != this.props.text) {
                this.finishEditing();
            }
        }
    }, {
        key: 'render',
        value: function render() {
            if (this.props.isDisabled) {
                var Element = this.props.element || this.props.staticElement;
                return _react2.default.createElement(
                    Element,
                    {
                        className: this.props.className,
                        style: this.props.style },
                    this.state.text || this.props.placeholder
                );
            } else if (!this.state.editing) {
                var Element = this.props.element || this.props.staticElement;
                return _react2.default.createElement(
                    Element,
                    {
                        className: this.props.className,
                        onClick: this.startEditing,
                        tabIndex: this.props.tabIndex,
                        style: this.props.style },
                    this.state.text || this.props.placeholder
                );
            } else {
                var Element = this.props.element || this.props.editingElement;
                return _react2.default.createElement(Element, {
                    onClick: this.clickWhenEditing,
                    onKeyDown: this.keyDown,
                    onBlur: this.finishEditing,
                    className: this.props.activeClassName,
                    placeholder: this.props.placeholder,
                    defaultValue: this.state.text,
                    onChange: this.textChanged,
                    style: this.props.style,
                    ref: 'input' });
            }
        }
    }]);

    return InlineEdit;
}(_react2.default.Component);

InlineEdit.propTypes = {
    text: _propTypes2.default.string.isRequired,
    paramName: _propTypes2.default.string.isRequired,
    change: _propTypes2.default.func.isRequired,
    placeholder: _propTypes2.default.string,
    className: _propTypes2.default.string,
    activeClassName: _propTypes2.default.string,
    minLength: _propTypes2.default.number,
    maxLength: _propTypes2.default.number,
    validate: _propTypes2.default.func,
    style: _propTypes2.default.object,
    editingElement: _propTypes2.default.string,
    staticElement: _propTypes2.default.string,
    tabIndex: _propTypes2.default.number,
    isDisabled: _propTypes2.default.bool,
    editing: _propTypes2.default.bool
};
InlineEdit.defaultProps = {
    minLength: 1,
    maxLength: 256,
    editingElement: 'input',
    staticElement: 'span',
    tabIndex: 0,
    isDisabled: false,
    editing: false
};
exports.default = InlineEdit;
